Add plan upgrade flow

dsas opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Part of DOTDASH-454

## Proposed Changes

Add a new super simple stepper flow that just exists to allow people to upgrade their plans in a distraction-free way.

In future we might want to test adding a domain step.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

We should unify all upgrade entry points into a single, full-screen PlansGrid because it creates a consistent and focused upgrade experience that reduces friction and confusion for users. Instead of pulling people out of context into the dashboard navigation, we can direct them to a clear, distraction-free surface that’s designed solely to help them compare plans and make a decision.

See also p58i-nge-p2

This change is looking to get the user interface right within the existing system, it's not looking at "hosting dashboardifying" the internals of how it works.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

<img width="1090" height="1253" alt="Screenshot 2025-10-08 at 18 37 31" src="https://github.com/user-attachments/assets/96229d3c-19ce-4097-9f98-100b9715f9b2" />


1. Go to /setup/plan-upgrade/plans?siteSlug=:siteSlug&redirect_to=/themes/:siteSlug



### Happy Path
3. You should see a plans grid
4. You should be able to select a plan
5. You should be able to increase the payment interval, or keep it the same, but not decrease it.
6. You should be able to complete checkout
7. You should be redirected to the url specified in `redirect_to` GET var.

### Other scenarios

1. **Authentication**: Anonymous user gets redirected to login
2. **Authorization**: User without access to :siteSlug gets treated appropriately
3. **Redirect**: Post-checkout redirect works with various URLs
4. **Plan Restrictions**: Verify unified-plans step shows only upgrade options (not downgrades) for existing sites
5. **Feature Restrictions**: Verify that the unified-plans step shows only upgrade options that provide the said feature if a feature GET var is supplieed, e.g. `&feature=sftp-and-database-access` should mean that only the current plan + business and e-commerce is listed.
6. Check the other onboarding flows to ensure their plans grids continue to work as before
8. Check /plans/:siteSlug to ensure that it continues to work as before.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

dsas committed:
WIP: Add plan upgrade flow

This is a flow that just exists to allow people to upgrade their plans
in a distraction-free way.

In future we might want to test adding a domain step.

dsas commented on client/landing/stepper/declarative-flow/flows/plan-upgrade/README.md:
```diff
@@ -0,0 +1,114 @@
+# Plan Upgrade Flow
```
I need to trim this down, it contains to do lists and other stuff too.

dsas committed:
Use intents system to limit the plans shown to current plan + higher plans

dsas committed:
Show the current plan as "Your plan"

dsas committed:
Add subheader

dsas committed:
Change display of upgrade credits

Make it look like the figma for the plans-upgrade intent - it displays
at the top-right corner, and isn't dismissible.

dsas committed:
Address race condition when fetching caps

dsas committed:
Use more appropriate page title

dsas committed:
Use constant for the plan-upgrade flow

dsas committed:
Fix issue with interval selector not working

Drill the isStepperUpgradeFlow prop down, so that the plan-upgrade
stepper flow generates the correct query-string URL format that
getIntervalType util function can parse.

dsas committed:
Limit plans in 'plan-upgrade' flow depending upon feature

Setting a feature= GET var when accessing the plan-upgrade flow, will
limit the plans displayed further to just the users plan + the plans
that provide the feature that are upgrades to the users current plan.

E.g. if supplying the `sftp-and-database-access` feature then only
display the current plan + atomic-capable plans.

dsas committed:
Simplify readme

dsas commented on client/landing/stepper/declarative-flow/flows/plan-upgrade/plan-upgrade.ts:
```diff
@@ -0,0 +1,177 @@
+		};
+	},
+
+	useAssertConditions(): AssertConditionResult {
```
This method is marked as deprecated, but every flow I looked at has similar logic here. 🤔

alshakero replied:
You can assert whatever you want in the initialize function. It can be async.

dsas commented on client/landing/stepper/declarative-flow/flows/plan-upgrade/plan-upgrade.ts:
```diff
@@ -0,0 +1,177 @@
+
+		useEffect( () => {
+			if ( ( siteSlug || siteId ) && ! site ) {
+				// Set a timeout - if no site after 5 seconds, give up
```
This is icky, but I’m not sure how else to identify which someone has given a siteSlug that they don't have access to. 

Without this I had race conditions between fetching the `site` data completing and checking permissions.

alshakero replied:
You can have these assertions in `initialize`.

alshakero commented on client/landing/stepper/declarative-flow/flows/plan-upgrade/plan-upgrade.ts:
```diff
@@ -0,0 +1,177 @@
+			result = {
+				state: AssertConditionState.FAILURE,
+				message: 'User must be logged in to access plan upgrade flow.',
+			};
```
You don't need this because `stepsWithRequiredLogin` takes care of this.

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/unified-plans/index.tsx:
```diff
@@ -84,6 +87,12 @@ type ProvidedDependencies = {
 const PlansStepAdaptor: StepType< {
 	submits: ProvidedDependencies;
```
You can add `accepts` here, adjacent to submits to add accepted props.
```suggestion
	submits: ProvidedDependencies;
	accepts: {
		isInSignup?: boolean;
		isStepperUpgradeFlow?: boolean;
		selectedFeature?: string;
	};
```

alshakero commented on client/landing/stepper/declarative-flow/internals/steps-repository/unified-plans/index.tsx:
```diff
@@ -84,6 +87,12 @@ type ProvidedDependencies = {
 	submits: ProvidedDependencies;
 } > = ( props ) => {
+	// Extract step-specific props from the flow configuration
+	const { isInSignup, isStepperUpgradeFlow, selectedFeature } = props as typeof props & {
```
Then you won't need this casting.

alshakero reviewed (commented):
Pretty neat! Uses the newest Stepper APIs 🚀  Left some comments here and there.

dsas committed:
Move logic from useAssertConditions to initialize

useAssertConditions was already deprecated, moving code to initalize is
the right thing to do.

This has required changing types - initialize was checking for false
before, but it wasn't a possible return value, now it is.

The login check has been removed because stepsWithRequiredLogin already
takes care of it.

Also standardised on url query var rather than param.

dsas committed:
Add accepts to unified plans step

This removes the need to cast props.

dsas committed:
Fix ts error

alshakero commented on client/landing/stepper/declarative-flow/internals/types.ts:
```diff
@@ -234,7 +236,14 @@ export type Flow = {
+	reduxStore: Store
+) => StepperStep[] | false | Promise< StepperStep[] | false >;
+
+/**
+ * Helper type to extract the steps array from the initialize return type, excluding false
+ */
+type ExtractSteps< T > = Exclude< Awaited< T >, false >;
```
Beautiful. Thanks! I didn't have a chance to implement this, but `initialize` is already documented to supposedly support this. So this is perfect.

alshakero commented on client/landing/stepper/declarative-flow/flows/plan-upgrade/plan-upgrade.ts:
```diff
@@ -0,0 +1,124 @@
+							const checkoutUrl = `/checkout/${ encodeURIComponent( siteSlug ) }/${ selectedPlan }`;
+
+							// Add redirect_to parameter if provided
+							const queryParams = new URLSearchParams();
```
You probably want to a `cancel_to` param set to `window.location.href.replace( window.location.origin, '' );` This way clicking the X in Checkout brings you back to the flow.

alshakero reviewed (approved):
I enjoyed reviewing this so much. This is probably the cleanest flow we have in Stepper. Left one commnet.

dsas committed:
Add back handling

Allow the checkout page to redirect back to the plans page.

Rather than have the plans page have a 'back' that goes back to the
checkout page (without adding a plan). Add a back that goes to a
location from the url, or back to /sites if not.

chriskmnds commented on packages/plans-grid-next/src/hooks/data-store/use-grid-plans.tsx:
```diff
@@ -51,39 +53,57 @@ const isGridPlanVisible = ( {
 	} = {},
 	isDisplayingPlansNeededForFeature,
-	planSlug,
+	gridPlanSlug,
```
I guess we didn't really need to change this :-)

dsas replied:
We didn't need to change it, no. I found it a little bit confusing having both `planSlug` and the `currentPlanSlug` I added, so it was a (perhaps unsuccessful) attempt to clarify it.

chriskmnds commented on packages/plans-grid-next/src/hooks/data-store/use-grid-plans.tsx:
```diff
@@ -168,6 +188,33 @@ export const usePlanTypesWithIntent = ( {
 			];
 			break;
+		case 'plans-upgrade': {
+			// Show current plan plus all higher-tier plans (upgrade options only)
```
This sounds like a behavior to abstract, as it's not the first time we implement this sort of thing (to only show upgrade options). 

Just a drive-by comment.
